From 7e3a6047c9a4d4ce888c93b495d941693577faab Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Thu, 9 Feb 2023 06:27:26 +0000
Subject: [PATCH] [Huawei] iommu: Add back function that support dm domain

Offering:HULK
CVE:NA
Reference:NA
Type:bugfix
DTS/AR:DTS2023021005454
reason: Add support for create direct mapping domain

Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
---
 drivers/iommu/iommu.c | 60 +++++++++++++++++++++++++++++++++++++++++++
 include/linux/iommu.h |  6 +++++
 2 files changed, 66 insertions(+)

diff --git a/drivers/iommu/iommu.c b/drivers/iommu/iommu.c
index f80596080..4241685a2 100644
--- a/drivers/iommu/iommu.c
+++ b/drivers/iommu/iommu.c
@@ -3181,6 +3181,66 @@ struct iommu_resv_region *iommu_alloc_resv_region(phys_addr_t start,
 }
 EXPORT_SYMBOL_GPL(iommu_alloc_resv_region);
 
+static int
+request_default_domain_for_dev(struct device *dev, unsigned long type)
+{
+	struct iommu_domain *domain;
+	struct iommu_group *group;
+	int ret;
+
+	/* Device must already be in a group before calling this function */
+	group = iommu_group_get(dev);
+	if (!group)
+		return -EINVAL;
+
+	mutex_lock(&group->mutex);
+
+	ret = 0;
+	if (group->default_domain && group->default_domain->type == type)
+		goto out;
+
+	/* Don't change mappings of existing devices */
+	ret = -EBUSY;
+	if (iommu_group_device_count(group) != 1)
+		goto out;
+
+	ret = -ENOMEM;
+	domain = __iommu_domain_alloc(dev->bus, type);
+	if (!domain)
+		goto out;
+
+	/* Attach the device to the domain */
+	ret = __iommu_attach_group(domain, group);
+	if (ret) {
+	       iommu_domain_free(domain);
+		goto out;
+	}
+
+	/* Make the domain the default for this group */
+	if (group->default_domain)
+		iommu_domain_free(group->default_domain);
+	group->default_domain = domain;
+
+	iommu_create_device_direct_mappings(group, dev);
+
+	dev_info(dev, "Using iommu %s mapping\n",
+		type == IOMMU_DOMAIN_DMA ? "dma" : "direct");
+
+	ret = 0;
+out:
+	mutex_unlock(&group->mutex);
+	iommu_group_put(group);
+
+	return ret;
+}
+
+/* Request that a device is direct mapped by the IOMMU */
+int iommu_request_dm_for_dev(struct device *dev)
+{
+	return request_default_domain_for_dev(dev, IOMMU_DOMAIN_IDENTITY);
+}
+EXPORT_SYMBOL_GPL(iommu_request_dm_for_dev);
+
 void iommu_set_default_passthrough(bool cmd_line)
 {
 	if (cmd_line)
diff --git a/include/linux/iommu.h b/include/linux/iommu.h
index 0f7d29ab9..75b74989a 100644
--- a/include/linux/iommu.h
+++ b/include/linux/iommu.h
@@ -539,6 +539,7 @@ extern void iommu_get_resv_regions(struct device *dev, struct list_head *list);
 extern void iommu_put_resv_regions(struct device *dev, struct list_head *list);
 extern void generic_iommu_put_resv_regions(struct device *dev,
 					   struct list_head *list);
+extern int iommu_request_dm_for_dev(struct device *dev);
 extern void iommu_set_default_passthrough(bool cmd_line);
 extern void iommu_set_default_translated(bool cmd_line);
 extern bool iommu_default_passthrough(void);
@@ -877,6 +878,11 @@ static inline int iommu_get_group_resv_regions(struct iommu_group *group,
 	return -ENODEV;
 }
 
+static inline int iommu_request_dm_for_dev(struct device *dev)
+{
+	return -ENODEV;
+}
+
 static inline void iommu_set_default_passthrough(bool cmd_line)
 {
 }
-- 
2.17.1

